README.md
# QUALIFIER: Quality Assessment of Gated Flow Cytometry Data
Mike Jiang, Greg Finak  
07/20/2015  

## Background
  The current `flowQ` package does the quality assessment on the ungated FCM data. However, there is a need to
  identify deviant samples by monitoring the consistencies of the underlying statistical properties of different gated cell 
  populations (such as white blood cells, lymphacytes, monocytes etc). The `flowQ` package was also not designed for dealing with
  large datasets. To meet these needs, We developed `QUALIFIER` package using the gating template created in flowJo and performing
  QA checks on different gated populations. It divides the data preprocessing from the actual outlier detection process so that the statistics
  are calcuated all at once and the outlier detection and visualization can be done more efficiently and interactively.
  `ncdfFlow` is used to solve the memory limit issue for large datasets. 





## Parsing the QA gating template
`parseWorkspace` function from `flowWorkspace` package is used to parse the flowJo workspace.
Since we are only interested in the gating strategy (gates defined in the template),`execute` can be set as FALSE  
to skip the actual gating process. `useInternal` is set as TRUE to use internal structure (c++) for faster parsing. 
Please see the documentation of `flowWorkspace` package for more details regarding how to use `parseWorkspace` function. 

```r
ws<-openWorkspace("./data/QA_MFI_RBC_bounary_eventsV3.xml")##should replace with your own xml workspace file path
GT<-parseWorkspace(ws,execute=FALSE,useInternal=TRUE)
```

## Apply the gating template to the data
Then gating hierarchy `gh_template` containing the actual template gates is extracted
from the result `GatingSet` "`GT`". 

```r
gh_template<-GT[[1]]  				
```

The `GatingSet` method here is the contructor that takes a `gatingHierarchy` object as the template and a list of FCS file names (`filenames`) that need the quality assurance.
The result "G" is the `GatingSet` that contains the gated data and some of the cell population statistics that can be viewed by `getPopStats` method.

```r
##datapath is the path where FCS files stores
G<-GatingSet(gh_template,filenames,path="./data")
getPopStats(G[[1]])
```
Optionally, `isNcdf` can be set to TRUE to support netCDF storage of large flow datasets when memory resource is limited. 

## Calculating the statistics
After flow data is gated, the statistics of the gated data need to be extracted and saved before the QA checks. 
`db` is an environment that serves as a container storing all the QA data and results.
Firsly, `initDB` function initializes and creates the data structures for QA. 
Then `qaPreprocess` is a convenient wrapper that calls underlining routines (`getQAStats},`saveToDB`)
to calculates/extracts statistics of gated cell populations and save them along with the gating set and the FCS meta data.
`metaFile` is a csv file that contains the extra annoations about each FCS file.It should at least have one column
speficifying FCS file names.


```r
db<-new.env()
initDB(db)
qaPreprocess(db=db,gs=G
			,metaFile="./data/FCS_File_mapping.csv" #should replace with your own FCS meta data file path
			,fcs.colname="FCS_Files"
			,date.colname=c("RecdDt","AnalysisDt")
	)
```
`fcs.colname` and `date.colname` are the arguments to identify the columns in the meta data (the flat table stored as a csv file)
that specifies the FCS filenames and the dates (to be formatted as to "m/d/y"). 
Once the preprocessing steps are finished, the data are ready for quality assessments. 

## Defining qaTasks 


`read.qaTask` reads external csv spreadsheet that contains the descriptions of each QA task and creates a list of `qaTask` objects. 


```r
checkListFile<-file.path(system.file("data",package="QUALIFIER"),"qaCheckList.csv.gz")
qaTask.list<-read.qaTask(db,checkListFile=checkListFile)
```

```
## Outlier function is not specified!
## 'outlier.norm' will be used as default outlier function.
## Outlier function is not specified!
## 'qoutlier' will be used as default outlier function.
```

```
## [1] "7 qaTask created ahd saved in db!"
```

```r
qaTask.list[1:2]
```

```
## $MFIOverTime
## qaTask: MFIOverTime 
## Level : Assay 
## Description : Fluorescence stability over time 
## population:  MFI
## Default formula :MFI ~ RecdDt | channel * stain
## <environment: 0x79594b8>
## Plot type:  xyplot
## $horiz
## [1] FALSE
## 
## $scales
## $scales$format
## [1] "%m/%d/%y"
## 
## 
## 
## $NumberOfEvents
## qaTask: NumberOfEvents 
## Level : Tube 
## Description : Number of Events Collected 
## population:  root
## Default formula :count ~ RecdDt | Tube
## <environment: 0x79e25c0>
## Plot type:  xyplot
## $horiz
## [1] FALSE
## 
## $scales
## $scales$format
## [1] "%m/%d/%y"
```
The `qaTask` can also be created individually by the contructor `makeQaTask`.

## Quality assessment and visualiztion 
`qaCheck` and `plot` are the two main methods to perform the quality assessment and visualize the QA results.
They both use the information stored in `qaTask` object 
and the `formula`, which is given either explicitly by the argument or implicitly by the `qaTask` object.
It is generally of the form y ~ x | g1 * g2 * ... , where y is the statistics to be checked in this QA and must be one of the four types:
			 
				"MFI": 
					Median Fluorescence Intensity of the cell population specified by `qaTask`,
			
				"proportion": 
						the percentage of the cell population specified by `qaTask` in the parent population, 
			
				"count": 
						the number of events of the cell population specified by `qaTask`,
			
				"spike": 
						the variance of intensity over time of each channel, which indicates the stability of the fluorescence intensity.

x specifies the variable plotted on x-axis (such as date) in the `plot` method.

g1, g2, ... are the conditioning variables, which divide the data into subgroups and 
apply the outlier detection whitin each individual group or plot the subgroups in different panels.
They may also be omitted, in which case the outlier detection is peformed in the entire dataset.

For example, RBC Lysis efficiency (percentage of WBC population) check is defined by `qaTask` .

```r
qaTask.list[["RBCLysis"]]
```

```
## qaTask: RBCLysis 
## Level : Tube 
## Description : Sufficient RBC lysis 
## population:  WBC_perct
## Default formula :proportion ~ RecdDt | Tube
## <environment: 0x7c288e0>
## Plot type:  xyplot
## $horiz
## [1] FALSE
## 
## $scales
## $scales$format
## [1] "%m/%d/%y"
```

According to the formula stored in `qaTask`, it uses the statistical property "proportion" and groups 
the data by "Tube" (or staining panel). "RecdDt" is only for the plotting purpose (specifing the x-axis). Cell population is defined as "WBC\_perct"

```r
qaCheck(qaTask.list[["RBCLysis"]]
       ,outlierfunc=list(func = outlier.cutoff
                        ,args =  list(lBound=0.8)
                       )
      )
```

`qaCheck` reads all the necessary information about the gated data from `qaTask` object.
Users only needs to specificy how the outliers are called. This is done by 
providing an outlier detection function `outlierfunc` that takes a numeric vector as input and returns a logical vector as the output.
Here `outlier.cutoff` provided by the package is used and threshold "lBound" is specified ("less than", use uBound for "larger than"). 

After the outliers are called, the results can be plotted by `plot` method.

```r
plot(qaTask.list[["RBCLysis"]],xlab="Record Date",ylab="percent")
```

![](QUALIFIER_files/figure-html/plot-RBCLysis-1.png) 

By default all the data are plotted, argument "subset" can be used to visualize a small subset.
 

```r
plot(qaTask.list[["RBCLysis"]],subset=Tube=='CD8/CD25/CD4/CD3/CD62L',xlab="Record Date",ylab="percent")
```

![](QUALIFIER_files/figure-html/plot-RBCLysis-subset-1.png) 

`clearCheck` is the method to removes the outlier results detected by the previous `qaCheck` call for the specific qaTask.

```r
clearCheck(qaTask.list[["RBCLysis"]])
```

With `scatterPlot` flag set as true and  `subset` properly specified plot method can generate scatter plots for the selected FCS files,

```r
plot(qaTask.list[["RBCLysis"]],subset=name=='06087181_F01_I010.fcs',scatterPlot=TRUE)
```
![QUALIFIER-plot-subset2](vignettes/QUALIFIER-plot-subset2.png)			

x term in the formula is normally ignored in `qaCheck`. However, when "plotType" of the 
`qaTask` is "bwplot", it is used as the conditioning variable that divides the data into subgroups,
 within which the `outlierfunc` is applied.   

```r
qaTask.list[["MNC"]]
```

```
## qaTask: MNC 
## Level : Assay 
## Description : Consistency of Lymphocyte/MNC Gate 
## population:  MNC
## Default formula :proportion ~ coresampleid
## <environment: 0x7b35798>
## Plot type:  bwplot
## $horiz
## [1] FALSE
## 
## $scales
## $scales$format
## [1] "%m/%d/%y"
```

This qaTask detects the significant variance of MNC cell population percentage among aliquots, which have the same "coresampleid".
Plot type of this object tells the method to group data by "coresampleid".   

```r
qaCheck(qaTask.list[["MNC"]],z.cutoff=1.5)
```

Interquartile Range based outlier detection function is used to detect outliers


```r
plot(qaTask.list[["MNC"]],proportion~factor(coresampleid),xlab="Sample ID",ylab="percent",scales=list(x=list(rot=45)))
```

![](QUALIFIER_files/figure-html/plot-MNC-1.png) 

The red circles in the boxplot indicate the possible outlier samples and the box of red color indicates the entire sample group
has significant variance and is marked as the group outlier. By default `qaCheck` uses normal-distribution-based outlier function to detect group outliers.  
User-defined function can also supplied through `gOutlierfunc` argument.Again the function should take a numeric vector as input and returns a logical vector as the output.
The formula supplied here in the plot method overwrites the one stored in the qaTask object, thus change the way of viewing the data.

With `scatterPlot` and `subset` arguments, scatter plots can be generated for the selected FCS files or sample groups,

```r
plot(qaTask.list[["MNC"]]
		,scatterPlot=TRUE
		,subset=coresampleid==11730)
```

![QUALIFIER-plot-MNC-scatter](vignettes/QUALIFIER-plot-MNC-scatter.png)	
		
We can also apply simple aggregation to the statisics through the formula.

```r
qaTask.list[["BoundaryEvents"]]
```

```
## qaTask: BoundaryEvents 
## Level : Channel 
## Description : Off-scale Boundary Events 
## population:  margin
## Default formula :proportion ~ RecdDt | channel
## <environment: 0x7ebcf08>
## Plot type:  xyplot
## $horiz
## [1] FALSE
## 
## $scales
## $scales$format
## [1] "%m/%d/%y"
```

Here the default formula only extracts the "proportion" from each individual channel.
In order to check the total percentage of boundary events of all channels for each fcs file,
we can write a new formula by applying aggregation function "sum" to "proportion" and group the data
by fcs file ("name" in this case).


```r
qaCheck(qaTask.list[["BoundaryEvents"]]
        ,sum(proportion) ~ RecdDt | name
        ,outlierfunc=list(func = outlier.cutoff
                          ,args =list(uBound=0.0003)
                          )
        )
```

And we still can visualize the results chanel by chanel.

```r
plot(qaTask.list[["BoundaryEvents"]],proportion ~ RecdDt | channel,xlab="Record Date",ylab="percent")
```

![](QUALIFIER_files/figure-html/plot-BoundaryEvents-1.png) 

 
Another three examples:
QA check of Fluorescence stability overtime using t-distribution based outlier detection function.     

```r
qaCheck(qaTask.list[["MFIOverTime"]]
		,rFunc= list (func = rlm
		              ,args = list(z.cutoff=3)
                  )
		)

plot(qaTask.list[["MFIOverTime"]]
		,y=MFI~RecdDt|stain
		,subset=channel%in%c('FITC-A')
		,rFunc=rlm
		,scales=list(y=c(relation="free"))
		,xlab="Record Date"
	)
```

![](QUALIFIER_files/figure-html/qaCheck-MFIOverTime-1.png) 

Note that the robust linear regression is applied in each group in order to capture the significant MFI change (highlighted in red color,if pvalues<0.05) over time.
The individual outliers within each group are detected based on the residue.
All the lattice options (like `scales` here in this example) can be passed to control the appearance of trellis plot,see the documentation of `lattice` for more details.


```r
qaCheck(qaTask.list[["spike"]]
			,outlierfunc = list (func = outlier.t
                                ,args = list(alpha=0.00001)
                                ) 
                
			)

plot(qaTask.list[["spike"]],y=spike~RecdDt|channel
		,subset=Tube=='CD8/CD25/CD4/CD3/CD62L'&channel%in%c('FITC-A')
		,xlab="Record Date"
	)
```

![](QUALIFIER_files/figure-html/plot-spike-1.png) 

When monitoring the total number of events for each tube, a pre-determined events number can be provided as the threshold to the qaCheck method.

`tubesEvents` pass to the argument of `qaCheck` could be a one-column data frame or a named list/vector. 
Threshold values are stored in the column or list/vecor and conditioning values stored in rownames or names of the list/vector.

```r
tubesEvents
```

```
##                             events
## CD8/CD25/CD4/CD3/CD62L       50000
## CD1c/IgD/CD27/CD19/IgM      150000
## HLADR/CD80/CD27/CD19/CD86   150000
## CD57/CD56/CD8/CD3/CD14       20000
## 6B11/Va24/CD8/BLK/CD4        50000
## CD8/CD69/CD4/CD3/HLADR       20000
## IgG1/IgG1/IgG1/IgG1/IgG1     20000
## Auto/Auto/Auto/Auto/Auto     20000
## CD11c/CD80/DUMP/HLADr/CD123 400000
## CD11c/CD86/DUMP/HLADr/CD123 400000
## LD/LD/LD/LD/LD               20000
## VD1/VD2/GD/BLK/CD3          250000
```


```r
qaCheck(qaTask.list[["NumberOfEvents"]]
		,formula=count ~ RecdDt | Tube
		,outlierfunc = list(func = outlier.cutoff
		                    ,args = list(lBound = 0.8 * tubesEvents)
                        )
		  )
```

```
## [1] "done!"
```

```r
plot(qaTask.list[["NumberOfEvents"]]
		,subset=Tube=='CD8/CD25/CD4/CD3/CD62L'
		,xlab="Record Date"
		,ylab="Cell counts"
	)
```

![](QUALIFIER_files/figure-html/plot-NumberOfEvents-1.png) 



```r
qaCheck(qaTask.list[["RedundantStain"]]
        ,gOutlierfunc = list(func = outlier.norm
                            , args =list(z.cutoff = 1)
                            )
         )
plot(qaTask.list[["RedundantStain"]]
		,y=proportion~factor(coresampleid)|channel:stain
		,subset=stain%in%c('CD8')
		,scales=list(x=list(cex=0.5,rot=45))
		,xlab="Sample ID"
		,ylab="percent"
	)
```

![](QUALIFIER_files/figure-html/plot-RedundantStain-1.png) 

## Creating quality assessment report
Besides the interactive visualization provided by `plot` method,we also provide one routine to generate  
all plots in one report.This function reads the QA results calculated by `qaCheck` and the meta information of each QA task provided in spreadsheet qaCheckList and 
generate the summary tables and svg plots. Svg plots provide tooltips containing the detail information about each sample as and 
hyperlinks of densityplot for each individual FCS file.  


```r
qaReport(qaTask.list
		,outDir="~/temp"
		,plotAll=FALSE
		,subset=as.POSIXlt(RecdDt)$year==(2007-1900)
		)
```

`plotAll` is the argument to control the plotting of the individual scatter plot for each FCS file. If TRUE, scatter plots for all the FCS files are generated. 
If FALSE, only the FCS files marked as outliers will be plotted. When it is set to "none", no scatter plot will be generated, which is helpful to provide a quick preview of the html report.
`subset` is the filter to only generate the report on a subset of flow data.

Note that if there is need to adjust the QA plot for each indivdiual qaTask in the report, the arguments must be stored in qaTask before qaReport method is called.
If there is no outlier is detected, the qaTask is not plotted by default. In order to change this setting, `htmlReport` can be set as TRUE to plot the qaTask regardlessly.      
This is particually useful for the task that tracks the longitudinal trend instead of the individual outliers.
Besides the lattice arguments that can be configured by `qpar` (for the summary xyplot and bwplot) and `scatterPar` (for the individual FCS xyplot).
`xlog`,`ylog` flag can be set for scatter plot to transform the flow data to log scale for the proper display of the stained channels.

```r
htmlReport(qaTask.list[["MFIOverTime"]])<-TRUE
rFunc(qaTask.list[["MFIOverTime"]])<-rlm
scatterPar(qaTask.list[["MFIOverTime"]])<-list(xlog=TRUE)
scatterPar(qaTask.list[["BoundaryEvents"]])<-list(xlog=TRUE)
scatterPar(qaTask.list[["RedundantStain"]])<-list(xlog=TRUE)
qpar(qaTask.list[["RedundantStain"]])<-list(scales=list(x=list(relation="free")))
```


## Conclusion
By the formula-based `qaCheck` and `plot` methods, different QA tasks 
can be defined and performed in a generic way.
And `plot` only reads the outlier detection results pre-calculated by `qaCheck`,
which reduces the cost of interactive visualization.

Two kinds of lattice plots are currently supported:xyplot and bwplot(boxplot), depending on the `plotType` in `qaTask` object.
When the output path is provided by `dest`, the svg plot is generated. 
In svg plot, each dot or box (or only the one marked as outliers) is annotated by the tooltip 
or hyperlink, which further points to the individual density plot of the gated population.




inst/NEWS
1.0.0

FEATURES:
	- outlier detection for different population stats(counts,proportion,MFI,Spike) of gated/ungated flow data
	- xyplot and boxplot on different dimensions
	- scatter plot for gated population
	- HTML report with svg support
	- fuzzy match of population name to select the same population from the different gating path
	
KNOWN ISSUES:
	
	- extracting populations stats needs further optimization on speed 
	- extend the formula parser to allow more functions to be applied by formula
	 
		 

R/qaPlot.R
# TODO: Add comment
# 
# Author: mike
###############################################################################
.FileNameGen<-function(prefix=NA,ID=NA,population=NA,channel=NA,stain=NA,stats=NA,...)
{
	fileName<-c(prefix,ID,basename(population),channel,stain,stats,...)
	fileName<-fileName[!is.na(fileName)]
	paste(fileName,collapse="_")
	
}
#TODO:to merge this single plot to groupplot since groupPlot is also able to produce the same xyplot
#just need to add timeline plot to groupPlot routine.
##single plot for each FCS
qa.singlePlot<-function(db,yy,statsType)
{
#	browser()
#	for(i in 1:nrow(yy))
#	{
#		curRow<-yy[i,]
	curRow<-yy
	ghInd<-which(getSamples(db$G)==curRow$name)
	if(length(ghInd)>0)
	{
		curGh<-db$G[[ghInd]]
		parentNode<-getParent(curGh,as.character(curRow$node))
		if(length(parentNode)>0)
		{
			x<-getData(curGh,parentNode)
		}else
		{
			x<-getData(curGh)
		}
		
		curGate<-getGate(curGh,as.character(curRow$node))
		
		#		curGate<-getGate(G[[which(getSamples(G)==curRow$name)]],as.character(curRow$node))
		#				browser()
		individualPlot(x,curGate,curRow,statsType)
	}
#	}	
}
individualPlot<-function(x,curGate,curRow,statsType)
{
	cols <- colorRampPalette(IDPcolorRamp(21,
					t(col2hsv(c("blue","green","yellow","red"))),
					fr=c(0.7,0)))
	
	#get gate
#	statsType<-curRow$stat
	channel<-as.character(curRow$channel)
	pop<-as.character(curRow$population)

#	curID<-pData(x)$id
	params<-flowCore:::colnames(x)

	if(is.object(curGate))
	{
		chnls<-parameters(curGate)
		fres <- filter(x,curGate)
	}else
	{
		chnls<-ifelse(statsType=="spike",channel,NA)
		fres<-NULL
	}
	
	

	if(statsType=="count"&&pop=="/root")##total cell count
	{
		##individual xyplot without gate
		xyplot(eval(parse(text=paste("`",params[2],"`~`",params[1],"`",sep="")))
						,data=x,smooth=FALSE,colramp=cols)
	}else
	{

		
		if(length(chnls)==1)
		{
			
			
			if(statsType=="spike")
			{
				t1<-paste("`",chnls[1],"`~`",params[grep("time",params,ignore.case=T)],"`",sep="")					
			}else
			{
				t1<-paste("`",params[2],"`~`",chnls[1],"`",sep="")
			}
		}else
		{
			t1<-paste("`",chnls[2],"`~`",chnls[1],"`",sep="")	
		}
		
		if(is.object(curGate))
		{
			gateName<-strsplit(curGate@filterId,split="\\.")[[1]][2]##remove prefix			
			mainTitle<-paste(gateName,"Gate")	
		}else
		{
			mainTitle<-""
		}
		
#			browser()	

		xyplot(x=as.formula(t1)
						,data=x
#							xlim=range(exprs(x[[1]])[,parameters(curGate)[1]]),
#							ylim=range(exprs(x[[1]])[,parameters(curGate)[2]]),
						,smooth=FALSE
						,colramp=cols
						,filter=fres
						,names=FALSE
#									pd=pData(x),
						,main=mainTitle
						,par.settings=list(gate.text=list(text=0.7
												,alpha=1
												,cex=1
												,font=1)
										,gate=list(
												fill="transparent"
												,lwd<-2
												,lty="solid"
												,alpha=1
												,col="red"
											)
										,flow.symbol=list(cex=ifelse(statsType=="spike",2,flowViz.par.get("flow.symbol")$cex))
								)
						,panel=panel.xyplot.flowframeEx
				)

	}

}
#TODO:check why colnames from flowCore is not dispatched correctly without namespace explicitly specified
##group plot for each sampleID or other aggregation ID
qa.GroupPlot<-function(db,yy)
{
	cols <- colorRampPalette(IDPcolorRamp(21,
					t(col2hsv(c("blue","green","yellow","red"))),
					fr=c(0.7,0)))

	pop<-unique(as.character(yy$population))
	if(length(pop)>1)
		stop("not the same population in lattice group plot!")
#	browser()
	#extract flowFrame and gate from each gating hierarchy
	frlist<-apply(yy,1,function(curRow){
#		browser()
		#get the parent population for the scatter plot
		
		curSampleInd<-which(getSamples(db$G)%in%curRow["name"])
		curGh<-db$G[[curSampleInd]]
		curNode<-as.character(curRow["node"])
		curGate<-getGate(curGh,curNode)
		parentNode<-getParent(curGh,curNode)
		parentNodeInd<-which(getNodes(curGh)%in%parentNode)
		if(length(parentNodeInd)>0)
		{
			fr<-getData(curGh,parentNodeInd)
		}else
		{
			fr<-getData(curGh)
		}
		list(frame=fr,gate=curGate)
	})
	names(frlist)<-yy[,"name"]
	
	#merge frames into flowSet for flowViz plot
	fs1<-flowSet(lapply(frlist,"[[","frame"))
	#append outlier flags
	if(!"outlier"%in%colnames(yy))
		yy$outlier<-FALSE
	pData(fs1)$outlier<-yy$outlier
	varMetadata(fs1)["outlier",]<-"outlier"

	#extract gates from the list
	gates<-lapply(frlist,"[[","gate")
	
	obj<-NULL
	if(!pop=="/root")##total cell count
	{
		
	
		fres<-filter(fs1,gates)
		if(length(parameters(fres[[1]]))==2)
		{
			t1<-paste("`",parameters(gates[[1]])[1],"`~`",parameters(gates[[1]])[2],"`",sep="")
		}else
		{
			t1<-paste("`",flowCore::colnames(fs1)[grep("SSC",flowCore::colnames(fs1))],"`~`",parameters(fres[[1]])[1],"`",sep="")
		}

		obj<-xyplot(x=as.formula(t1),
						data=fs1,
						smooth=FALSE,
						colramp=cols,
						filter=fres,
						names=FALSE,
						pd=pData(fs1),
						par.settings=list(gate.text=list(text=0.7
														,alpha=1
														,cex=1
														,lineheight=2
														,font=1)
											,gate=list(
													fill="transparent"
													,lwd<-2
													,lty="solid"
													,alpha=1
													,col="red"
													)
											),
						panel=panel.xyplot.flowsetEx
				)
	}
	
			
			
	
	return(obj)
	
#	plotObj
}





setMethod("plot", signature=c(x="qaTask"),
		function(x,y,...){

			#assign null to formula if it is missing
			if(missing(y))
				y<-formula(x)
			plot.qaTask(qaObj=x,formula=y,...)
		})

plot.qaTask<-function(qaObj,formula,subset,pop,width,height,par,isTerminal=TRUE,fixed=FALSE,dest=NULL,rFunc=NULL,plotAll=FALSE,scatterPlot=FALSE)
{
#	browser()
	par_old<-qpar(qaObj)
	if(!missing(par))##overwrite the elements of par slot of qa object if provided by argument
	{
		for(x in names(par))
			eval(substitute(par_old$v<-par$v,list(v=x)))
				
		qpar(qaObj)<-par_old
	}
#	browser()
	if(missing(width))
		width<-QUALIFIER:::width(qaObj)
	if(missing(height))
		height<-QUALIFIER:::height(qaObj)
	lattice.options(print.function=plot.trellisEx)
#	browser()
	db<-getData(qaObj)
	##query db
	curGroup<-NULL
#	browser()
	
	if(is.null(qpar(qaObj)$horiz))
		qpar(qaObj)$horiz<-FALSE


	#parse the formula
	formuRes<-.formulaParser(formula)
	#decide the statsType(currently only one of the terms can be statType,we want to extend both in the future)
	
	statsType<-matchStatType(db,formuRes)
	if(missing(pop))
		pop<-getPop(qaObj)
#		browser()
	
	if(missing(subset))
		yy<-queryStats(db,statsType=statsType,pop=pop,isTerminal=isTerminal,fixed=fixed)
	else
		yy<-queryStats(db,statsType=statsType,substitute(subset),pop=pop,isTerminal=isTerminal,fixed=fixed)
	if(nrow(yy)==0)
	{
		message("no samples are matched!")
		return()
		
	}
	
#	browser()
	#check if the conditioning variable is of factor type
	for(curGroupBy in formuRes$groupBy)
	{

		curCol<-substitute(yy$v,list(v=curGroupBy))
		if(!class(eval(curCol))=="factor")
		{
			eval(substitute(v<-as.factor(v),list(v=curCol)))
		}
	}
#	browser()
	if(getName(qaObj)=="BoundaryEvents")
		yy<-base::subset(yy,value>min(yy$value))##filter out those zero-value records which may cause slow plotting
			

	#append the outlier flag
	yy$outlier<-yy$sid%in%base::subset(db$outlierResult,qaID==qaID(qaObj))$sid
	yy$gOutlier<-yy$sid%in%base::subset(db$GroupOutlierResult,qaID==qaID(qaObj))$sid
	
	#reshape the data to include the column of the statType which can be passed to lattice	as it is
	yy<-as.data.frame(cast(yy,...~stats))

#	dest=list(...)$dest

	if(!is.null(dest))
	{	
		sfile<-tempfile(pattern=getName(qaObj),tmpdir=dest,fileext=".svg")
#		browser()
		devSVGTips(sfile,width=width,height=height)
		isSvg<-TRUE
	}else
	{
		isSvg<-FALSE
	}
	
	
#	browser()
	plotObjs=new.env()
	if(scatterPlot)
	{
		##if scatterPlot flag is true then just plot the scatter plot
		
		if(statsType=="spike"||(statsType=="count"&&yy$population[1]=="/root"))
		{
#			browser()
			thisCall<-lapply(1:nrow(yy),function(i)
			{
#						browser()
						qa.singlePlot(db,yy[i,,drop=FALSE],statsType)
			})
			
			
		}else
		{
			thisCall<-qa.GroupPlot(db,yy)
		}
		
		
	}else
	{#otherwise, plot the summary plot (either xyplot or bwplot)

		par<-qpar(qaObj)
		par$subscripts<-TRUE
		par$strip<-TRUE
		
		xlab<-par$xlab
		ylab<-par$ylab
		main<-par$main
		pch<-par$pch
		layout<-par$layout
		cex<-par$cex
		par.strip.text<-par$par.strip.text
		scales<-par$scales
		xscale.components<-par$xscale.components
		if(plotType(qaObj)=="xyplot")
		{
			##parse the viz par
			if(is.null(main))
				par$main<-paste(description(qaObj),curGroup,sep=":")	
			if(is.null(pch))
				par$pch<-19		
			if(is.null(par.strip.text))
				par$par.strip.text<-list(lines=2)	
			if(is.null(scales))
				par$scales<-list(x=c(cex=0.7
						#						,rot=45	
											))
#			if(is.null(xscale.components))
#				par$xscale.components<-function(...) {
#					ans <- xscale.components.default(...)
#					ans$bottom$ticks$at<-seq(from=min(yy$RecdDt),to=max(yy$RecdDt),by="2 month")
#					ans$bottom$labels$at<-seq(from=min(yy$RecdDt),to=max(yy$RecdDt),by="2 month")
#					ans$bottom$labels$labels <- zoo::as.yearmon(seq(from=min(yy$RecdDt),to=max(yy$RecdDt),by="2 month"))
#					ans
#				}
			thisCall<-quote(
							xyplot(x=formula,data=yy
									,groups=outlier
									,panel=function(x=x,y=y,data=yy,dest.=dest,plotObjs.=plotObjs,plotAll.=plotAll,statsType.=statsType,...){
										panel.xyplotEx(x,y,data=data,dest.=dest,plotObjs.=plotObjs,plotAll.=plotAll,statsType.=statsType,db=db,...)
								#if regression function is supplied, then plot the regression line
										if(!is.null(rFunc))
										{
	#							browser()
										
											reg.res<-try(rFunc(y~x),silent=TRUE)
											if(all(class(reg.res)!="try-error"))
											{
												sumry<-summary(reg.res)
												if(class(sumry)=="summary.rlm"){
													coefs<-coef(sumry)
													t.value<-coefs[,"t value"]
													slope<-coefs[2,"Value"]
													intercept<-coefs[1,"Value"]
													df<-summary(reg.res)$df
													pvalues<-pt(abs(t.value),df=df[1],lower.tail=FALSE)
													intercept.p<-pvalues[1]
													slope.p<-pvalues[2]
												}else if (class(sumry)=="summary.lm"){
													pvalues<-coefficients(sumry)[,4]
													slope<-coefficients(sumry)[2,1]
													intercept.p<-pvalues[1]
													slope.p<-pvalues[2]
													
												}	
												if(any(pvalues<0.05))
												{
													regLine.col<-"red"
												}else
												{
													regLine.col<-"black"
												}
												curVp<-current.viewport()
												
												
												#							panel.lines(y=rFunc(y~x)$fitted,x=x,type="l",col="black",lty="solid")
												panel.text(x=mean(curVp$xscale)
														,y=quantile(curVp$yscale)[4]
														,labels=paste("s=",format(slope*30,digits=2)
																#														," v=",format(var(y),digits=2)
																,"\np=",paste(format(slope.p,digits=2),collapse=",")
														)
														,cex=0.5
												#										,col="white"		
												)
												
												panel.abline(reg.res,col=regLine.col,lty="dashed")
												}
										}
								
									}
								)
							)
		}
		
		if(plotType(qaObj)=="bwplot")
		{
			
			plot.symbol<-trellis.par.get("plot.symbol")
			plot.symbol$col<-"red"
			trellis.par.set("plot.symbol",plot.symbol)
			
		lattice.options(print.function=plot.trellisEx)
	
			if(qpar(qaObj)$horiz)
			{
				groupBy.Panel<-as.character(formuRes$yTerm)#formula[[3]][[2]])
				
			}else
			{
				groupBy.Panel<-as.character(formuRes$xTerm)
		}

			
			
			
	
#			if(is.null(xlab))
#				par$xlab<-groupBy.Panel
#			if(is.null(ylab))
#				par$ylab<-statsType
			if(is.null(main))
				par$main<-paste(description(qaObj),curGroup,sep=":")	
			if(is.null(pch))
				par$pch<-"."	
			if(is.null(cex))
				par$cex<-5
			if(is.null(par.strip.text))
				par$par.strip.text<-list(lines=2)	
			if(is.null(scales))
				par$scales<-list(x=c(cex=0.9
											,rot=45	
											)
										)
									
			thisCall<-quote(bwplot(x=formula,data=yy
									,groupBy=groupBy.Panel
									,panel=function(data=yy,dest.=dest,plotObjs.=plotObjs
														,plotAll.=plotAll
														,statsType.=statsType
														,...){
#										browser()
											panel.bwplotEx(data.=data,dest.=dest
															,plotObjs.=plotObjs,plotAll.=plotAll
															,statsType.=statsType
															,db=db,...)
											}
									)
							)
			
		}
		#append the par list
		thisCall<-as.call(c(as.list(thisCall),par))
#		browser()
		thisCall<-eval(thisCall)
#		print(thisCall)
	}
	
#	browser()
	if(isSvg)
	{
		print(thisCall)
		ret<-dev.off()
		message("Saving to ",sfile)


		fileNames<-ls(plotObjs)
		
		if(!is.null(fileNames)&&plotAll!="none")
		{
#			browser()
#			t1<-Sys.time()
			for(dfile in fileNames)
			{
#				browser()
				curPlotObj<-get(dfile,plotObjs)
				#determine the pic size by the number of pannels
				nPanels<-length(curPlotObj$packet.sizes)
				wRatio<-ceiling(sqrt(nPanels))
				hRatio<-floor(sqrt(nPanels))
				png(file.path(dest,"individual",dfile),width=300*wRatio,height=300*hRatio)
				print(curPlotObj)
				dev.off()
				
			}
			
#			t2<-Sys.time()
			message("individual plots genetated.")
			
		}
		
#		browser()
		.postProcessSVG(sfile)
			
		return(basename(sfile))
	}else
	{
		thisCall
	}
	
	
	
}

